Build a multi-statement query when inserting fixtures

Edouard-chin opened the pull request:
Build a multi-statement query when inserting fixtures:

- The `insert_fixtures` method can be optimized by making a single multi statement query for all fixtures having the same connection instead of doing a single query per table
  - The previous code was bulk inserting fixtures for a single table, making X query for X fixture files
  - This patch builds a single **multi statement query** for every tables. Given a set of 3 fixtures (authors, dogs, computers):
  ```ruby
    # before
    %w(authors dogs computers).each do |table|
      sql = build_sql(table)
      connection.query(sql)
    end

    # after

    sql = build_sql(authors, dogs, computers)
    connection.query(sql)
  ```
- `insert_fixtures` is now deprecated, `insert_fixtures_set` is the new way to go with performance improvement
- My tests were done with an app having more than 700 fixtures, the time it takes to insert all of them was around 15s. Using a single multi statement query, it took on average of 8 seconds
- In order for a multi statement to be executed, mysql needs to be connected with the `MULTI_STATEMENTS` [flag](https://dev.mysql.com/doc/refman/5.7/en/c-api-multiple-queries.html), which is done before inserting the fixtures by reconnecting to da the database with the flag declared. Reconnecting to the database creates some caveats:
  1. We loose all open transactions; Inside the original code, when inserting fixtures, a transaction is open. Multple delete statements are [executed](https://github.com/rails/rails/blob/a681eaf22955734c142609961a6d71746cfa0583/activerecord/lib/active_record/fixtures.rb#L566) and finally the fixtures are inserted. The problem with this patch is that we need to open the transaction only after we reconnect to the DB otherwise reconnecting drops the open transaction which doesn't commit all delete statements and inserting fixtures doesn't work since we duplicated them (Primary key duplicate exception)...
    - In order to fix this problem, the transaction is now open directly inside the `insert_fixtures` method, right after we reconnect to the db
    - As an effect, since the transaction is open inside the `insert_fixtures` method, the DELETE statements need to be executed here since the transaction is open later
  2. The same problem happens for the `disable_referential_integrity` since we reconnect, the `FOREIGN_KEY_CHECKS` is reset to the original value
    - Same solution as 1. , the disable_referential_integrity can be called after we reconnect to the transaction
  3. When the multi statement query is executed, no other queries can be performed until we paginate over the set of results, otherwise mysql throws a "Commands out of sync" [Ref](https://dev.mysql.com/doc/refman/5.7/en/commands-out-of-sync.html)
    - Iterating over the set of results until `mysql_client.next_result` is false. [Ref](https://github.com/brianmario/mysql2#multiple-result-sets)
- Removed the `active_record.sql "Fixture delete"` notification, the delete statements are now inside the INSERT's one
- On mysql the `max_allowed_packet` is looked up:
  1. Before executing the multi-statements query, we check the packet length of each statements, if the packet is bigger than the max_allowed_packet config, an `ActiveRecordError` is raised
  2. Otherwise we concatenate the current sql statement into the previous and so on until the packet is `< max_allowed_packet`

Edouard-chin committed:
Build a multi-statement query when inserting fixtures:

- The `insert_fixtures` method can be optimized by making a single multi statement query for all fixtures having the same connection instead of doing a single query per table
  - The previous code was bulk inserting fixtures for a single table, making X query for X fixture files
  - This patch builds a single **multi statement query** for every tables. Given a set of 3 fixtures (authors, dogs, computers):
  ```ruby
    # before
    %w(authors dogs computers).each do |table|
      sql = build_sql(table)
      connection.query(sql)
    end

    # after

    sql = build_sql(authors, dogs, computers)
    connection.query(sql)
  ```
- `insert_fixtures` is now deprecated, `insert_fixtures_set` is the new way to go with performance improvement
- My tests were done with an app having more than 700 fixtures, the time it takes to insert all of them was around 15s. Using a single multi statement query, it took on average of 8 seconds
- In order for a multi statement to be executed, mysql needs to be connected with the `MULTI_STATEMENTS` [flag](https://dev.mysql.com/doc/refman/5.7/en/c-api-multiple-queries.html), which is done before inserting the fixtures by reconnecting to da the database with the flag declared. Reconnecting to the database creates some caveats:
  1. We loose all open transactions; Inside the original code, when inserting fixtures, a transaction is open. Multple delete statements are [executed](https://github.com/rails/rails/blob/a681eaf22955734c142609961a6d71746cfa0583/activerecord/lib/active_record/fixtures.rb#L566) and finally the fixtures are inserted. The problem with this patch is that we need to open the transaction only after we reconnect to the DB otherwise reconnecting drops the open transaction which doesn't commit all delete statements and inserting fixtures doesn't work since we duplicated them (Primary key duplicate exception)...
    - In order to fix this problem, the transaction is now open directly inside the `insert_fixtures` method, right after we reconnect to the db
    - As an effect, since the transaction is open inside the `insert_fixtures` method, the DELETE statements need to be executed here since the transaction is open later
  2. The same problem happens for the `disable_referential_integrity` since we reconnect, the `FOREIGN_KEY_CHECKS` is reset to the original value
    - Same solution as 1. , the disable_referential_integrity can be called after we reconnect to the transaction
  3. When the multi statement query is executed, no other queries can be performed until we paginate over the set of results, otherwise mysql throws a "Commands out of sync" [Ref](https://dev.mysql.com/doc/refman/5.7/en/commands-out-of-sync.html)
    - Iterating over the set of results until `mysql_client.next_result` is false. [Ref](https://github.com/brianmario/mysql2#multiple-result-sets)
- Removed the `active_record.sql "Fixture delete"` notification, the delete statements are now inside the INSERT's one
- On mysql the `max_allowed_packet` is looked up:
  1. Before executing the multi-statements query, we check the packet length of each statements, if the packet is bigger than the max_allowed_packet config, an `ActiveRecordError` is raised
  2. Otherwise we concatenate the current sql statement into the previous and so on until the packet is `< max_allowed_packet`

sgrif reviewed (approved):
I'd like to see the commit message and PR description updated to make it more clear that this is moving from one query per table to one query per table sent as a single string. This generally looks good to me, though.

Only other point of note is that it's a shame that this has so much git churn. Perhaps we can keep the existing `insert_fixtures` definition as a separate method, which we call in a loop so there's less rightward drift and git churn?

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -529,12 +529,27 @@ def default_index_type?(index) # :nodoc:
+      def insert_fixtures(fixture_set, tables_to_delete = [])
+        iterate_over_results = -> { while raw_connection.next_result; end; }
+
+        without_sql_mode("NO_AUTO_VALUE_ON_ZERO") do
```
This changes a session variable...

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -529,12 +529,27 @@ def default_index_type?(index) # :nodoc:
-      def insert_fixtures(*)
-        without_sql_mode("NO_AUTO_VALUE_ON_ZERO") { super }
+      def insert_fixtures(fixture_set, tables_to_delete = [])
+        iterate_over_results = -> { while raw_connection.next_result; end; }
```
Do no other adapters need something like this?

It doesn't feel like great layering for `insert_fixtures` to be touching `raw_connection` at all, really.

sgrif replied:
No, MySQL is the only backend which forces you to care about what's in the network buffers, and manually flush it to avoid errors.

matthewd replied:
Okay, cool. As I'm paranoid, I think I'd like to have a test that does a compound fixture load with an insert that succeeds followed by one that fails (inserting null into a not null column, say), and confirms we get the exception we're expecting. i.e., that all drivers [including out-of-tree] not only receive a compound result set without stuffing up the connection, but actually notice if a later result contains an error.

sgrif replied:
For reference, the behavior that most other backends provide is that all queries are executed, but only the return value of the final query is provided. This is almost always handled at the client level, it's rare for the server to actually accept semicolon-delimited SQL strings (MySQL is the only one in this case AFAIK).

I agree we should have the test you've described though.

Edouard-chin replied:
Ack, will take care of writing the test. Thanks for the explanations

Edouard-chin replied:
Done in https://github.com/rails/rails/pull/31422/commits/7f8b345022887d1b7df3f486f9637a4aa7be1941

matthewd commented on activerecord/lib/active_record/fixtures.rb:
```diff
@@ -540,47 +540,46 @@ def self.create_fixtures(fixtures_directory, fixture_set_names, class_names = {}
-                end
-                deleted_tables[conn] << table
-              end
+        fixture_sets_by_connection.each_value do |set|
```
`each` would give us `conn` for free

Edouard-chin replied:
The `fixture_sets_by_connection` was original grouped by connection_id (now that I think of the variable name was confusing).

My original reasoning was that I didn't know how expensive it was to group by the connection itself. I probably overthought this though

matthewd commented on activerecord/lib/active_record/fixtures.rb:
```diff
@@ -540,47 +540,46 @@ def self.create_fixtures(fixtures_directory, fixture_set_names, class_names = {}
-                conn.reset_pk_sequence!(fs.table_name)
+            table_rows.each do |table, rows|
+              unless deleted_tables[conn].include? table
+                deleted_tables[conn] << table
```
Is `deleted_tables[conn]` just `table_rows_for_connection.keys`?

Edouard-chin replied:
Yes good catch !

matthewd commented on activerecord/test/cases/fixtures_test.rb:
```diff
@@ -64,7 +64,9 @@ def initialize
 
     def call(_, _, _, _, values)
-      @events << values[:sql] if values[:sql] =~ /INSERT/
+      return unless values[:name] == "Fixtures Insert"
```
I think the conditional insertion is a bit easier to read than the double-negative of an early exit on such a simple method.

(Why the change of condition?)

matthewd commented on activerecord/test/cases/fixtures_test.rb:
```diff
@@ -73,8 +75,10 @@ def test_bulk_insert
-        create_fixtures("bulbs")
+        create_fixtures("bulbs", "authors", "computers")
+
         assert_equal 1, subscriber.events.size, "It takes one INSERT query to insert two fixtures"
```
If we're going to borrow an existing test, let's keep its contents in sync.

Edouard-chin commented:
Thanks for the quick feedback @sgrif ! I split my PR into 2 commits. The first one makes all change in order for the feature to work. The second is cosmetic and remove all trailing spaces that we don't need since they were part of a inner block that is now removed.

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -384,7 +384,30 @@ def insert_fixtures(fixtures, table_name)
+      # INSERT INTO table2 (col1, col2) VALUES (?, ?), (?, DEFAULT);\n
+      #
+      # Overridden in adapters that require something beyond a simple insert (eg. Oracle).
+      def insert_fixtures(fixture_set, tables_to_delete = [])
```
In line with @sgrif's comment: this is public API.

Edouard-chin replied:
Ah right, then I guess I don't have a choice but to split this method. This would have the downside of creating another git churn :(

```ruby
def build_sql
  returns the sql string
end

def insert_fixtures
  build_sql
  execute
end

def insert_multiple_fixtures # new method to generate a single multi query statement
  total_sql = ''.dup

  fixtures.each do
    total_sql << build_sql
  end

  execute
end
```

rafaelfranca replied:
I think we will need to do it. Call the new method that does the multi insert `insert_fixtures_set`.

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -384,7 +384,30 @@ def insert_fixtures(fixtures, table_name)
+      #
+      # Overridden in adapters that require something beyond a simple insert (eg. Oracle).
+      def insert_fixtures(fixture_set, tables_to_delete = [])
+        total_sql = String.new
```
`"".dup` -- both because it's the prevailing style, and for the encoding

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -384,7 +384,30 @@ def insert_fixtures(fixtures, table_name)
+
+        fixture_set.each do |table_name, fixtures|
+          total_sql << build_sql_for_fixtures_insert(fixtures, table_name)
+          total_sql << ";\n"
```
Do we need to consider max_allowed_packet and/or similar settings for other backends?

Edouard-chin commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -384,7 +384,30 @@ def insert_fixtures(fixtures, table_name)
+
+        disable_referential_integrity do
+          transaction(requires_new: true) do
+            tables_to_delete.each { |table| delete "DELETE FROM #{quote_table_name(table)}", "Fixture Delete" }
```
I don't know if people rely on the `sql.active_record` notification when fixtures are deleted, but otherwise the DELETE statements can also be part of the same query when inserting fixtures which would make things even faster

matthewd replied:
Is it not just as slow to run lots of deletes? (and PK resets, for that matter)

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -384,7 +384,30 @@ def insert_fixtures(fixtures, table_name)
+        end
+
+        disable_referential_integrity do
+          transaction(requires_new: true) do
```
Won't this break if there's a surrounding transaction?

I guess the reconnect would break such a transaction no matter what, actually. :confused:

Edouard-chin replied:
> I guess the reconnect would break such a transaction no matter what, actually.

Yes the reconnect would break any open transaction, is this something we should be concerned about?

matthewd commented on activerecord/test/cases/fixtures_test.rb:
```diff
@@ -73,8 +75,14 @@ def test_bulk_insert
+        create_fixtures("bulbs", "authors", "computers")
+
         assert_equal 1, subscriber.events.size, "It takes one INSERT query to insert two fixtures"
+        if current_adapter?(:Mysql2Adapter)
```
Beyond being ugly, we can't do this because other adapters use our test suite, and have their own name-quoting mechanisms.

Edouard-chin replied:
TIL, didn't know other adapters uses rails test suite. For my own curiosity could you point which one, I'm curious to see how they do to reuse the test suite

matthewd commented on activerecord/test/cases/fixtures_test.rb:
```diff
@@ -73,8 +75,14 @@ def test_bulk_insert
+
         assert_equal 1, subscriber.events.size, "It takes one INSERT query to insert two fixtures"
+        if current_adapter?(:Mysql2Adapter)
+          assert_match(/INSERT INTO `bulbs`.*INSERT INTO `authors`.*INSERT INTO `computers`/m, subscriber.events.first)
```
This loses [the assertion of] the previous bulk behaviour: that it all happens in a single INSERT per table, and not one per row -- this would pass even if we were sending a huge stream of single-row INSERTs.

Edouard-chin replied:
👍 , I decided to write a separate test instead

matthewd commented on activerecord/test/cases/fixtures_test.rb:
```diff
@@ -73,8 +75,14 @@ def test_bulk_insert
+        if current_adapter?(:Mysql2Adapter)
+          assert_match(/INSERT INTO `bulbs`.*INSERT INTO `authors`.*INSERT INTO `computers`/m, subscriber.events.first)
+        else
+          assert_match(/INSERT INTO "bulbs".*INSERT INTO "authors".*INSERT INTO "computers"/m, subscriber.events.first)
```
If we're going to go to this much effort, maybe we should also consider other fast-insert mechanisms. PostgreSQL is much faster with `COPY`, for example.

Edouard-chin commented:
Thanks for the great review @matthewd and @sgrif !

Outlining this since it got hidden by GH after my push but it's not resolved.

> If we're going to go to this much effort, maybe we should also consider other fast-insert mechanisms. PostgreSQL is much faster with COPY, for example.

Edouard-chin commented:
Hello friends and happy new year !

Trying to get back on this, there is 2 points that weren't yet addressed in my latest commit.

> If we're going to go to this much effort, maybe we should also consider other fast-insert mechanisms. PostgreSQL is much faster with COPY, for example.

Will be happy to work on this, just need confirmation this is something we want for PostgreSQL. Do you think needs to be done inside the same PR ?

> Do we need to consider max_allowed_packet and/or similar settings for other backends?

What do you think should be the appropriate behaviour if `ER_NET_PACKET_TOO_LARGE` is thrown? Shall we try to reconnect to the DB and try again with a smaller insert query, or just warn the user?

matthewd commented:
> Do you think [PostgreSQL COPY] needs to be done inside the same PR ?

No, that was really just me thinking aloud, as a generalization of the "fixture insertion is worth strongly optimizing" thesis that this PR implies.

> What do you think should be the appropriate behaviour if ER_NET_PACKET_TOO_LARGE is thrown? Shall we try to reconnect to the DB and try again with a smaller insert query, or just warn the user?

I'm not sure... it seems like a lot of bother for us to arrange a retry, but just telling the user that we can't insert their fixtures (while not really offering much they can do to solve the problem) doesn't seem terribly helpful, either. Are there a known default values for that limit? (e.g. the base MySQL default, plus any distro configured defaults across a handful of major Linuxes)

Might it be practical for us to proactively look up the limit, and then use it when combining the tables together? Outright giving up if a single table is too big feels perfectly fine to me, while applying a limit across the entire DB seems less so. And while it's more code than ignoring limits entirely, I suspect it'd be simpler/neater to always consider the limit than to change behaviour upon retry.

If it's reasonable to assume the limit is practically infinite, then I'm fine with just ignoring it, at least until we get reports of it causing problems for people. But this PR's premise is essentially that people have lots of tables; if (lots of tables) * (large-but-reasonable data set) is at risk of nudging the limit, it seems worth some consideration.

> This changes a session variable... -- https://github.com/rails/rails/pull/31422#discussion_r156524929

... and won't that be immediately reset by the reconnect in `with_multi_statements`?

Edouard-chin commented:
Thanks for the detailed comment, it helped me greatly understand what you had in mind!

> This changes a session variable... -- #31422 (comment)
> ... and won't that be immediately reset by the reconnect in with_multi_statements?

Good catch! You are right, this was immediately reset by the reconnect. Since no tests were broken I have opened another PR https://github.com/rails/rails/pull/31663 to remove it.
For now, I have modified this PR to reconnect first and then modify the session variable.

> Might it be practical for us to proactively look up the limit, and then use it when combining the tables together?

It's a good idea and I implemented it in https://github.com/rails/rails/pull/31422/commits/899ae7200462462b522600ffd95d7ef626074ff8 , it makes things way more easier than implementing a retry logic.

Last but not least I created a new `insert_fixtures_set` method in order to not touch the `insert_fixtures` one which was public API.

Edouard-chin commented:
Ah shoot one test is failing on CI but not locally due to privileges

> ActiveRecord::StatementInvalid: Mysql2::Error: Access denied; you need (at least one of) the SUPER privilege(s) for this operation: SET GLOBAL max_allowed_packet=16777216

sgrif commented:
Easiest solution is to grant super privileges to the rails user [here](https://github.com/rails/rails/blob/master/ci/travis.rb#L9)

sgrif commented:
Also just to document this somewhere public, for the record there is a limit on PG. It's a limit at the protocol level, 2^31 bytes. Shouldn't be a problem for this use case unless someone has gigs of binary data in their fixtures.

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -358,33 +358,28 @@ def insert_fixture(fixture, table_name)
+          transaction do
+            tables_to_delete.each { |table| delete "DELETE FROM #{quote_table_name(table)}", "Fixture Delete" }
+            packet_size = total_sql.bytesize
+            raise ActiveRecordError, "Fixtures set is too large (#{packet_size} bytes). Consider increasing the max_allowed_packet variable." if packet_too_large?(packet_size)
```
I was hoping we could do this check a few lines above, and turn `total_sql` into an array of `< max_allowed_packet`-sized chunks.

matthewd commented on activerecord/test/cases/fixtures_test.rb:
```diff
@@ -79,6 +79,56 @@ def test_bulk_insert
+      error = assert_raises(ActiveRecord::ActiveRecordError) { create_fixtures(*fixtures) }
+      assert_match(/Fixtures set is too large/, error.message)
+    ensure
+      ActiveRecord::Base.connection.execute("SET GLOBAL max_allowed_packet=#{old_max_allowed_packet}")
```
Does this take effect immediately, or only upon reconnection? I think I saw a suggestion it could be the latter somewhere. That's fine for the earlier one, because we already reconnect to do the insertion.. but might mean we need another reconnect here?

Alternatively, it may be simpler to create new fixture(s) bigger than the default limit (using erb; no need for big actual files)?

matthewd commented:
> Easiest solution is to grant super privileges to the rails user

I'd rather our test suite not require DB superpowers to run. If it's the best option, it'd be reasonable to do that for CI, but ensure the test just skips when running with less privileges... but all else being equal, it seems nice to have CI enforcing the "no privileged operations" rule.

> PG [has] a limit at the protocol level, 2^31 bytes

:+1: yep, that meets my definition of 'practically infinite' in this context. (I wouldn't be surprised if other DBs had very similar limitations at about that point... PG's is hardly the only protocol developed at a time when 32 bits was Big Enough.)

Edouard-chin commented:
> I was hoping we could do this check a few lines above, and turn total_sql into an array of < max_allowed_packet-sized chunks.

💯 , done in my latest commit

Split the total_sql into chunks:

- Proactively looking for the max_allowed_packet and split the insert statement into smaller chunks to avoid mysql throwing a ER_NET_PACKET_TOO_LARGE
- If a single table is too big, raising a preventive ActiveRecordError since we won't be able to execute the query anyway
- Decided to stub the max_allowed_packet inside tests in order to not have to deal with super privileges. Added a memoization on the `max_allowed_packet`, my rational being that it's unlikely this variable changes, but maybe I'm too naive /shrug

sgrif commented:
> I'd rather our test suite not require DB superpowers to run. If it's the best option, it'd be reasonable to do that for CI, but ensure the test just skips when running with less privileges...

The only other option is to leave the `max_packet_size` stuff on MySQL untested

sgrif commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -533,8 +533,43 @@ def insert_fixtures(*)
+        def with_multi_statements
+          previous_flags = @config[:flags]
+          @config[:flags] = Mysql2::Client::MULTI_STATEMENTS
+          reconnect!
```
Why does this require reconnecting? This should just require a call to `mysql_set_server_option` on the C side

Edouard-chin replied:
Nice, wasn't aware of `mysql_set_server_option`. I can't find any reference of it in mysql2 though :(

matthewd replied:
Yeah, looks like we might have to go with this for now, and then see about getting that method added to mysql2 in future.

(Maybe you can get it sneaked in to the [upcoming release](https://github.com/rails/rails/issues/31636).)

kamipo replied:
How about adding `MULTI_STATEMENTS` by default?
We already have `FOUND_ROWS` to make it the same as PostgreSQL's behavior. Related https://github.com/rails/rails/pull/25370#issuecomment-230157091.

https://github.com/rails/rails/blob/fd04c8cbb6627ef65ea728b5026b21bd24471c18/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb#L12-L20

sgrif replied:
I'd prefer not to have `MULTI_STATEMENTS` by default. It has a lot of gotchas on MySQL which can cause really subtle bugs

sgrif commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -533,8 +533,43 @@ def insert_fixtures(*)
+          yield
+        ensure
+          @config[:flags] = previous_flags
+          reconnect!
```
Still shouldn't require reconnecting. [I'll just leave this here](https://github.com/diesel-rs/diesel/blob/5d5e0685cb5392cb01dba1473ecb06d45db3351c/diesel/src/mysql/connection/raw.rs#L101-L117) (I assume mysql2 exposes that function somewhere)

sgrif reviewed (approved):
One comment. Looks fine to me once that's resolved, or made clear why it can't be resolved

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -358,33 +358,33 @@ def insert_fixture(fixture, table_name)
 
-        values = fixtures.map do |fixture|
-          fixture = fixture.stringify_keys
+        fixture_set.each do |table_name, fixtures|
```
How about something more like:

```ruby
fixture_inserts = fixture_set.map { |table_name, fixtures| build_fixture_sql(fixtures, table_name) }
total_sql = combine_multi_statements(fixture_inserts)
```

Where `combine_multi_statements` in abstract is `[statements.join]`, and all the details of *why* some statements may not be combinable (i.e., there are length limits) can be pushed down to the MySQL adapter.

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -358,33 +358,33 @@ def insert_fixture(fixture, table_name)
-              Arel.sql("DEFAULT")
+        disable_referential_integrity do
+          transaction do
+            tables_to_delete.each { |table| delete "DELETE FROM #{quote_table_name(table)}", "Fixture Delete" }
```
The above would also give us a handy helper to squish these together too

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -358,33 +358,33 @@ def insert_fixture(fixture, table_name)
       def insert_fixtures(fixtures, table_name)
```
We can't kill/change this because it's documented, but we can deprecate it. Especially as it is itself fairly new, I don't think we have any strong motivation to keep it long term.

Edouard-chin replied:
Done in my latest commit

matthewd commented on activerecord/lib/active_record/fixtures.rb:
```diff
@@ -540,47 +540,38 @@ def self.create_fixtures(fixtures_directory, fixture_set_names, class_names = {}
+              table_rows_for_connection[table].unshift(*rows)
             end
           end
+          conn.insert_fixtures_set(table_rows_for_connection, table_rows_for_connection.keys)
```
Should we just leave the `.keys` up to that method to handle?

Edouard-chin replied:
`conn.insert_fixtures_set(table_rows_for_connection, table_rows_for_connection)`

felt weird to me and thought of just doing `conn.insert_fixtures_set(table_rows_for_connection)` but
since `insert_fixtures_set` is public API, if one wants to insert fixtures without deleting any tables, he wouldn't be able to anymore. WDYT?

matthewd commented on activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb:
```diff
@@ -373,6 +373,18 @@ def insert_fixtures(rows, table_name)
 
+      def insert_fixtures_set(fixture_set, tables_to_delete = [])
+        disable_referential_integrity do
+          transaction(requires_new: true) do
```
Why is this `requires_new` but the other isn't?

Edouard-chin replied:
Good catch, my bad the other should have it too. It was extracted from here

https://github.com/rails/rails/blob/fd04c8cbb6627ef65ea728b5026b21bd24471c18/activerecord/lib/active_record/fixtures.rb#L558

Edouard-chin commented:
Addressed the suggestion/comments. Thanks again for the time spent reviewing this ❤️

kamipo commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -533,8 +533,57 @@ def insert_fixtures(*)
+        end
+
+        def max_allowed_packet
+          @max_allowed_packet ||= execute("SELECT @@SESSION.max_allowed_packet").first[0].to_i
```
How about using `@max_allowed_packet ||= show_variable("max_allowed_packet")`?

Edouard-chin replied:
very cool, wasn't aware of this method!

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -356,35 +356,32 @@ def insert_fixture(fixture, table_name)
-              with_yaml_fallback(bind.value_for_database)
-            else
-              Arel.sql("DEFAULT")
+        total_sql = Array.wrap(combine_multi_statements(fixture_inserts, tables_to_delete))
```
Sorry, that part was unclear:

I meant `insert_fixtures_set` can call `combine_multi_statements` a second time... or actually, even just combine them:

```ruby
table_deletes = tables_to_delete.map { |table| "DELETE FROM #{quote_table_name table}" }
total_sql = Array.wrap(combine_multi_statements(table_deletes + fixture_inserts))
```

Edouard-chin replied:
👍 it will make things even faster.  I was just unsure if people relied on the activerecord.sql notifications when deleting fixtures (https://github.com/rails/rails/pull/31422#discussion_r156528814).

matthewd replied:
Yeah, I think it's fine for those to change; arguably people are more likely to be interested in the inserts than the deletes, and we're already radically changing those -- and no-one complained last time.

(If it's going to be doing deletes *and* inserts, I guess we should change the label on the `execute` to `"Fixtures Load"` or something, too.)

matthewd commented on activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:
```diff
@@ -533,8 +533,57 @@ def insert_fixtures(*)
+          elsif previous_packet.nil?
+            false
+          else
+            (current_packet.bytesize + previous_packet.bytesize) > max_allowed_packet
```
Do we need to allow for any padding etc here? "Packet" sounds to me like it could be including a surrounding struct of some sort. I just tried to check this experimentally, and ended up just confusing myself: in the `mysql` CLI tool, after setting `max_allowed_packet` to 1024, I was able to `SELECT 'xxxx'` for 10 KB of 'x's :confused:

Edouard-chin replied:
I got confused as well and should have explained beforehand.

> The packet message buffer is initialized to net_buffer_length bytes, but can grow up to max_allowed_packet bytes when needed. This value by default is small, to catch large (possibly incorrect) packets.

Because the [net-buffer-length](https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_net_buffer_length) by default is set to 16384 bytes, setting the `max_allowed_packet` to 1024 has actually no effect. It will have effect only if `max_allowed_packet` is higher than the `net-buffer-length`, otherwise it doesn't need to shrink.

matthewd replied:
Ah, thanks. After setting them both to 1024, I was able to run queries up to length **1022**, plus a trailing semicolon and/or newline (omitting them didn't allow me to add another character, so I'm not counting them). I haven't dug in to work out whether that's a quirk of the `mysql` CLI (e.g., maybe it always adds a semicolon if one isn't present?)

That's certainly close enough that it clearly doesn't count any surrounding data... but can you perhaps experiment with this real code, and work out whether we need to give it a 1-2 byte buffer? I guess it's just a matter of starting with a pair of fixtures whose inserts sum to ~1030, confirming they get sent as two executes, then removing one character at a time down to ~1020 (which should be sent as a single multi), and ensuring that each one of those runs succeeds.

matthewd commented:
This is looking great! Thank *you* for the time spent working on it! :heart: :heart:

sgrif commented:
@matthewd Is there anything else you think is absolutely blocking this? We can fix the 1-2 byte margin of error in 5.2.1 IMO. The CI failure looked unrelated, but it was on AR:mysql2, which is definitely the thing that would be affected by this so I restarted the job. Are you good with merging once CI is green and we can fix the minor stuff that's left later?

@Edouard-chin Are you planning on wrapping up the changes to this today/tomorrow?

Edouard-chin commented:
> Are you planning on wrapping up the changes to this today/tomorrow?

@sgrif Absolutely ! I'll address the last comment this afternoon

Edouard-chin committed:
Combine delete and insert statements in the same query

matthewd commented:
@sgrif I'm worried that if we ship this with a yet-to-be-investigated possible 0.000006% chance of failure, we'll forget about it and it'll just [extremely occasionally] annoy people who get very unlucky. If it's hard to check, I'm fine with shipping as is.. but it seems like it should be pretty straight-forward to determine while we're here.

(Alternatively, we could just subtract two [or more] from the calculated length. That way the worst case scenario is that we possibly split something a few bytes too early; there's no real downside to underestimating the length available, compared to the possibility of an actual failure if we overestimate.)

Edouard-chin commented:
I have some bandwidth today. Let me confirm if the padding is needed

On Jan 23, 2018 10:40 AM, "Matthew Draper" <notifications@github.com> wrote:

> @sgrif <https://github.com/sgrif> I'm worried that if we ship this with a
> yet-to-be-investigated possible 0.000006% chance of failure, we'll forget
> about it and it'll just [extremely occasionally] annoy people who get very
> unlucky. If it's hard to check, I'm fine with shipping as is.. but it seems
> like it should be pretty straight-forward to determine while we're here.
>
> (Alternatively, we could just subtract two [or more] from the calculated
> length. That way the worst case scenario is that we possibly split
> something a few bytes too early; there's no real downside to
> underestimating the length available, compared to the possibility of an
> actual failure if we overestimate.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/31422#issuecomment-359829924>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHvvhuNoyCffwj_6M-aE2YBreCmaSWoJks5tNf1ugaJpZM4Q_tNj>
> .
>

Edouard-chin commented:
So you were right @matthewd, played around with this real code, and we need a 2 bytes margin. Taking care of it now

Edouard-chin committed:
Allow a 2 bytes margin:

- mysql will add a 2 bytes margin to the statement, so given a `max_allowed_packet` set to 1024 bytes, a 1024 bytes fixtures will no be inserted (mysql will throw an error)
- Preventing this by decreasing the max_allowed_packet by 2 bytes when doing the comparison with the actual statement size

matthewd commented:
Awesome, thanks so much for checking that!

:shipit: :shipit: :ship: :it:

Edouard-chin commented:
Amazing!  It was a fun feature to implement. Thanks all for your guidance

pdobb commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -356,35 +356,33 @@ def insert_fixture(fixture, table_name)
-              with_yaml_fallback(bind.value_for_database)
-            else
-              Arel.sql("DEFAULT")
+        table_deletes = tables_to_delete.map { |table| "DELETE FROM #{quote_table_name table}".dup }
```
Why the `.dup` on the brand new string literal?

kamipo replied:
Looks like `sql << ";\n"` for the line: https://github.com/rails/rails/pull/31422/files#diff-98ce5dc0cd01f8b0435ff6feac10d8a8R551

Edouard-chin replied:
Yeah Kamipo pointed the correct reason
